### PR TITLE
[4.0] Remove caption js

### PR DESCRIPTION
**Pull request by Bakual**
Pull Request to remove the legacy caption.js

### Summary of Changes
Removes the JHtml::_('behavior.caption') calls, the method itself and its JS files and test code.

### Testing Instructions
* Make sure caption.js isn't requested anymore in any pageview
* Captions should work as before because we actually moved to use `figure` and `figcaption` long ago. The JS was kept and loaded for B/C reasons with existing old articles.

### Expected result
Captions work as before

### Actual result
Captions work as before

### Documentation Changes Required
Mark the method as deprecated in 3.9 (I can do that PR) and in the docs.

### CODE REVIEW

**Comment by dgrammatiko** [comment]
We need to remove the related js test: https://github.com/joomla/test-javascript/tree/master/src/caption

I'll do a PR for that (but I still have to convert at least one basic script test there as an example, so give me some time for this...)

File: components/com_contact/tmpl/categories/default.php [hunk 1]
```diff
@@ -10,7 +10,6 @@
 defined('_JEXEC') or die;
 
 JHtml::addIncludePath(JPATH_COMPONENT . '/helpers');
-JHtml::_('behavior.caption');
 JHtml::_('behavior.core');
```
Comment (dgrammatiko) [inline components/com_contact/tmpl/categories/default.php]: do we still need the core here?

**Comment by brianteeman** [comment]
> Captions should work as before because we actually moved to use figure and figcaption long ago. The JS was kept and loaded for B/C reasons with existing old articles.

New articles created in 3.8 do not use figure and figcaption. So what is the plan to convert all articles on upgrade

**Comment by dgrammatiko** [comment]
@brianteeman are you referring to intro/full image or the images in the full text field (editor)?

**Comment by brianteeman** [comment]
intro/full is the only one that has caption

**Comment by dgrammatiko** [comment]
So we can update their jLayouts to use figcaption/figure tags

**Comment by dgrammatiko** [comment]
This I think should be ok
```php
<?php if (isset($images->image_fulltext) && !empty($images->image_fulltext)) : ?>
	<?php $imgfloat = empty($images->float_fulltext) ? $params->get('float_fulltext') : $images->float_fulltext; ?>
	<figure class="pull-<?php echo htmlspecialchars($imgfloat); ?> item-image">
		<img src="<?php echo htmlspecialchars($images->image_fulltext); ?>" alt="<?php echo htmlspecialchars($images->image_fulltext_alt); ?>" itemprop="image"/>
		<?php if ($images->image_fulltext_caption) :
			echo '<figcaption class="caption">'
					. htmlspecialchars($images->image_fulltext_caption)
				. '</figcaption>';
		endif; ?>
	</figure>
<?php endif; ?>
```

**Comment by Bakual** [comment]
@brianteeman Since J3.2.2 the image button in the editor should create figure and figcaption elements. My PR back then was https://github.com/joomla/joomla-cms/pull/2664 and when I tested it now it still works this way.
Of course, articles created before that time would have "broken" captions as long as you don't load a replacement JS (eg using a plugin) or edit the articles to use figure. There is no way we can do that on updates.
I assumed we had changed the into/fullimages layouts as well, but apparently that was never done? Of course that then needs to be done in this PR as well. No need to update anything on the content for this.

Or do I miss something?

**Comment by brianteeman** [comment]
> I assumed we had changed the into/fullimages layouts as well, but apparently that was never done? Of course that then needs to be done in this PR as well. No need to update anything on the content for this.

yes that is all that needs to be done

**Comment by dgrammatiko** [comment]
> My PR back then was #2664 and when I tested it now it still works this way.

Well it's not working like that anymore with the new media manager. 😓
We need to fix that ASAP

**Comment by Bakual** [comment]
Adjusted the layouts for the intro and full image to use `figure` and `figcaption` tags.
The `figure` is always added, regardless if there is a caption or not.

@brianteeman I've read that the `figure` element should have an aria `role=group` if there is a caption included. I've added that one. Maybe you can verify if that is correct.

File: components/com_contact/tmpl/categories/default.php (hunk 1 above)
Comment (Bakual) [reply components/com_contact/tmpl/categories/default.php]: I honestly have no clue why core is loaded there.

File: layouts/joomla/content/full_image.php [hunk 2]
```diff
@@ -9,15 +9,19 @@
...
 defined('JPATH_BASE') or die;
 
-$params = $displayData->params;
-
+$params  = $displayData->params;
+$images  = json_decode($displayData->images);
+$caption = $images->image_fulltext_caption;
 ?>
-<?php $images = json_decode($displayData->images); ?>
+<?php  ?>
```
Comment (Quy) [inline layouts/joomla/content/full_image.php]: Remove

Comment (Bakual) [reply layouts/joomla/content/full_image.php]: Done, thanks!

**Comment by brianteeman** [comment]
@bakual see https://developer.paciellogroup.com/blog/2011/08/html5-accessibility-chops-the-figure-and-figcaption-elements/

File: components/com_contact/tmpl/categories/default.php (hunk 1 above)
Comment (dgrammatiko) [reply components/com_contact/tmpl/categories/default.php]: ok we need it for the translated text

**Comment by Bakual** [comment]
@brianteeman I can't say I understand everything written there. My knowledge about accessibility is quite limited. I got the part about the role=group from a german site (https://wiki.selfhtml.org/wiki/HTML/Textstrukturierung/figure).
What I got from your article is that the caption should be before the image, and in fact with the caption.js currently it's like that as well. So I moved that around.

File: layouts/joomla/content/full_image.php [hunk 3]
```diff
@@ -9,15 +9,18 @@
...
-<?php endif; ?>
+	<figure class="pull-<?php echo htmlspecialchars($imgfloat, ENT_COMPAT, 'UTF-8'); ?> item-image"<?php echo ($caption) ? ' role="group"' : ''; ?>>
+		<?php if ($caption) : ?>
+			<figcaption class="caption"><?php echo htmlspecialchars($caption, ENT_COMPAT, 'UTF-8'); ?></figcaption>
+		<?php endif; ?>
+		<img src="<?php echo htmlspecialchars($images->image_fulltext, ENT_COMPAT, 'UTF-8'); ?>"
+			 alt="<?php echo htmlspecialchars($images->image_fulltext_alt, ENT_COMPAT, 'UTF-8'); ?>"
+			 itemprop="image"/>
+	</figure>
+<?php endif; ?>
```
Comment (laoneo) [inline layouts/joomla/content/full_image.php]: Needs a new line here.

**Comment by laoneo** [comment]
@brianteeman are the changes ok regarding accessibility?
@dgrammatiko can you fix the related js tests from the testing repo?
@Bakual can you fix the conflicts here that we can move to the next step?

**Comment by brianteeman** [comment]
better to ask the accessibility team than me

**Comment by Bakual** [comment]
Rebased PR and fixed the conflict

**Comment by chmst** [comment]
@Bakual goupin roles are used to identify related form controls, so I do not see a necessity in figure/figcaption

**Comment by Bakual** [comment]
Ok, I'll remove that part then. Makes code simpler :smile:

**Comment by laoneo** [comment]
Opened a pr https://github.com/joomla/test-javascript/pull/7 in the testing repo to remove the caption.js tests. Once this is merged, we can merge this one here too.

**Comment by laoneo** [comment]
Once #20449 is merged, the JS tests will always be fetched from the correct branch instead of the latest version.

**Comment by laoneo** [comment]
Great work, thanks @Bakual!

**Comment by Bakual** [comment]
Follow up PR to deprecate it in J3: https://github.com/joomla/joomla-cms/pull/20451

**Comment by infograf768** [comment]
Beware https://github.com/joomla/joomla-cms/pull/20503

**Comment by infograf768** [comment]
BTW, I do not understand what is behind the idea that the caption should be before the image.
My test show that with this template it displays above the image, which is a nonsense.

**Comment by chmst** [comment]
@infograf768 it is more a matter of taste but figcaption can be used above or below (or both).

**Comment by infograf768** [comment]
@chmst 
I understand it is possible to display the caption above or below with figcaption but, nevertheless, the most common usage is below.
Will make another PR after #20503 is merged.

**Comment by dgrammatiko** [comment]
!Idea: let's have some options in the backend for the placement of the caption...


PS. I'm joking 😂

**Pull request merged**